include/revng/BasicAnalyses/GeneratedCodeBasicInfo.h

```diff
@@ -250,41 +243,77 @@ class GeneratedCodeBasicInfo : public llvm::ModulePass {
 
   /// \brief Return the program counter of the next (i.e., fallthrough)
   ///        instruction of \p TheInstruction
-  MetaAddress getNextPC(llvm::Instruction *TheInstruction) const {
+  MetaAddress getNextPC(llvm::Instruction *TheInstruction) {
     auto Pair = getPC(TheInstruction);
     return Pair.first + Pair.second;
   }
 
-  llvm::CallInst *getFunctionCall(llvm::BasicBlock *BB) const {
+  static llvm::CallInst *getFunctionCall(llvm::BasicBlock *BB) {
     return getFunctionCall(BB->getTerminator());
   }
 
   // TODO: is this a duplication of FunctionCallIdentification::isCall?
   // TODO: we could unpack the information too
-  llvm::CallInst *getFunctionCall(llvm::Instruction *T) const {
+  static llvm::CallInst *getFunctionCall(llvm::Instruction *T) {
+    revng_assert(T != nullptr);
     revng_assert(T->isTerminator());
-    auto It = T->getIterator();
-    auto End = T->getParent()->begin();
-    while (It != End) {
-      It--;
-      if (llvm::CallInst *Call = getCallTo(&*It, "function_call"))
+    llvm::Instruction *Previous = getPrevious(T);
+    while (Previous != nullptr && isMarker(Previous)) {
+      if (auto *Call = getCallTo(Previous, "function_call"))
         return Call;
 
-      if (not isMarker(&*It))
-        return nullptr;
+      Previous = getPrevious(Previous);
     }
 
     return nullptr;
   }
 
-  bool isFunctionCall(llvm::BasicBlock *BB) const {
+  static bool isFunctionCall(llvm::BasicBlock *BB) {
     return isFunctionCall(BB->getTerminator());
   }
 
-  bool isFunctionCall(llvm::Instruction *T) const {
+  static bool isFunctionCall(llvm::Instruction *T) {
     return getFunctionCall(T) != nullptr;
   }
 
+  static llvm::BasicBlock *getSuccessor(llvm::BasicBlock *BB) {
+    return getSuccessor(BB->getTerminator());
+  }
+
+  /// \brief Return the successor basic block given a function_call marker.
+  static llvm::BasicBlock *getSuccessor(llvm::Instruction *T) {
+    if (auto *Call = getFunctionCall(T)) {
+      if (auto *Successor = llvm::dyn_cast<llvm::BlockAddress>(Call->getOperand(0)))
+        return Successor->getBasicBlock();
+    }
+
+    return nullptr;
+  }
+
+  static llvm::BasicBlock *getFallthrough(llvm::BasicBlock *BB) {
+    return getFallthrough(BB->getTerminator());
+  }
+
+  /// \brief Return the fall-through basic block given a function_call marker.
+  static llvm::BasicBlock *getFallthrough(llvm::Instruction *T) {
+    if (auto *Call = getFunctionCall(T)) {
+      auto *Fallthrough = llvm::cast<llvm::BlockAddress>(Call->getOperand(1));
+      return Fallthrough->getBasicBlock();
+    }
+
+    return nullptr;
+  }
```

Comment:
```
include/revng/FunctionCallIdentification/FunctionCallIdentification.h:  llvm::BasicBlock *getFallthrough(llvm::Instruction *T) const {
```

Reply:
Yeah, I do remember this. I think we can fix these redundant cases for good, considering also [this](https://github.com/revng/revng/blob/develop/include/revng/BasicAnalyses/GeneratedCodeBasicInfo.h#L262). Shall we keep this (which fits here) and remove `getFallthrough` from `FCI`?

Reply:
Works. We'll need to reorganize GCBI. What matters now is to reduce code duplication.